packages/cardano-services-client/src/ChainHistoryProvider/BlockfrostChainHistoryProvider.ts

```diff
@@ -475,11 +476,31 @@ export class BlockfrostChainHistoryProvider extends BlockfrostProvider implement
   }
 
   public async transactionsByHashes({ ids }: TransactionsByIdsArgs): Promise<Cardano.HydratedTx[]> {
-    try {
-      return Promise.all(ids.map((id) => this.fetchTransaction(id)));
-    } catch (error) {
-      throw this.toProviderError(error);
+    const cachedTransactions: Cardano.HydratedTx[] = [];
+    const idsToFetch: Cardano.TransactionId[] = [];
+
+    for (const id of ids) {
+      if (this.cache.has(id)) {
+        cachedTransactions.push(this.cache.get(id)!);
+      } else {
+        idsToFetch.push(id);
+      }
+    }
+
+    let fetchedTransactions: Cardano.HydratedTx[] = [];
+    if (idsToFetch.length > 0) {
+      try {
+        fetchedTransactions = await Promise.all(idsToFetch.map((id) => this.fetchTransaction(id)));
+
+        for (const tx of fetchedTransactions) {
+          this.cache.set(tx.id, tx);
+        }
+      } catch (error) {
+        throw this.toProviderError(error);
+      }
     }
+
+    return [...cachedTransactions, ...fetchedTransactions];
```

Comment:
Maybe it would be better to return in the same order from the function call.

Reply:
```
Promise.all(
ids.map(id => {
  if (this.cache.has(id)) return Promise.resolve(this.cache.get(id));
  try { return this.fetchTransaction(id)} catch (error) { throw this.toProviderError(error); }
}
)

Reply:
done here https://github.com/input-output-hk/cardano-js-sdk/pull/1569/commits/4dce93bdc29ffae9f6e0f3b3f173dce0dc52186f